### PR TITLE
[api] use ActiveJob for BackendPackage.refresh_dirty method

### PR DESCRIPTION
**Pull request by evanrolfe**
As part of https://trello.com/c/Tx9d3bUb/467-5-p4-event-system-change-delayed-jobs-to-be-created-with-activejobs

### CODE REVIEW

File: src/api/app/models/backend_package.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@ class BackendPackage < ApplicationRecord
   # this is called from the UpdatePackageMetaJob and clockwork
   def self.refresh_dirty
     Package.dirty_backend_package.pluck(:project_id).uniq.each do |project_id|
-      Project.find(project_id).delay(priority: 10).update_packages_if_dirty
+      Project::UpdatePackagesIfDirty.perform_later(project_id)
```
Comment (mdeniz) [inline src/api/app/models/backend_package.rb]: What happen with the priority? isn't necessary anymore?

File: src/api/app/jobs/project/update_packages_if_dirty.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class Project
```
Comment (mdeniz) [inline src/api/app/jobs/project/update_packages_if_dirty.rb]: Mmmm using Project here could be "dangerous" or misleading??
Isn't better something like ProjectJobs ?

File: src/api/app/jobs/project/update_packages_if_dirty.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+  class UpdatePackagesIfDirty < ApplicationJob
```
Comment (mdeniz) [inline src/api/app/jobs/project/update_packages_if_dirty.rb]: Shouldn't this class name end with **Job**??

File: src/api/app/jobs/project/update_packages_if_dirty.rb (hunk 2 above)
Comment (evanrolfe) [inline src/api/app/jobs/project/update_packages_if_dirty.rb]: In what way is it dangerous or misleading?

Comment (mdeniz) [inline src/api/app/jobs/project/update_packages_if_dirty.rb]: "Dangerous" because your are opening the Project class here, so everything that someone add here by mistake overwrites the model definition.

Misleading in the sense of that your are defining parts of the Project model class in other file...

Comment (evanrolfe) [inline src/api/app/jobs/project/update_packages_if_dirty.rb]: > "Dangerous" because your are opening the Project class here, so everything that someone add here by mistake overwrites the model definition.

No this doesn't overwrite the model definition? This is a nested class, there is no overriding going on.. 

> Misleading in the sense of that your are defining parts of the Project model class in other file...

This is a nested class. Thats different from defining a part of the `Project` class in another file.

Comment (mdeniz) [inline src/api/app/jobs/project/update_packages_if_dirty.rb]: Nested where? what is the namespace for it?
What if I come after you and I add something like this in this file:
```ruby
class Project

  def name
    "whatever"
  end
.... your actual code
end
```

Then I'm overriding the **name** method of the Project model

Comment (mdeniz) [inline src/api/app/jobs/project/update_packages_if_dirty.rb]: For me is better to not use the same Namespace (in this case is the Project class)

Comment (evanrolfe) [inline src/api/app/jobs/project/update_packages_if_dirty.rb]: > Nested where? what is the namespace for it?

Its a classes nested inside another class. A ruby class can either be nested inside a module or a class.

> What if I come after you and I add something like this in this file:
> Then I'm overriding the **name** method of the Project model

Then don't do that?

Comment (mdeniz) [inline src/api/app/jobs/project/update_packages_if_dirty.rb]: I know how classes/modules can be nested, I meant that the outer namespace is the Project model class not a different module or class. I though that you meant that Project::UpdatePackagesIfDirty was already inside any other module/class

Ok, anyway. If someone tries to add something wrong to this we should be alert 😄

**Comment by adrianschroeter** [comment]
IMHO the better way here is to have a single job which just fetches all missing BackendPackage items without the need to have events for them. There was some not merged implementation here: https://github.com/openSUSE/open-build-service/pull/2168

**Comment by evanrolfe** [comment]
@adrianschroeter in the last sprint henne,moises and i analysed three different solutions to the delayed job issues we were having and decided to do it this way. See this document for more details (your solution is under solution #3):
https://github.com/openSUSE/open-build-service/wiki/Event-Subsystem-Architecture-Review

We decided against doing it that way because it means that you effectively have another table on top of the delayed jobs table which is being using to keep track of the async work being done. This also requires more work to be done in order to keep track of backend packeges which fail to update, more work to implement retries, and more work to make sure that the same backend package is not worked on twice. Also the way these delayed jobs were done by using locks to prevent other delayed jobs from working on the same thing has caused many mysql deadlock errors.

So if delayed job is having problems processing one event per backed package, I think we should look at why that is, if we can fix it or if we need to use a different backend for active job. Rather than adding more work arounds to the codebase in order to deal with delayed job not scaling.

File: src/api/app/jobs/project/update_packages_if_dirty.rb (hunk 3 above)
Comment (evanrolfe) [inline src/api/app/jobs/project/update_packages_if_dirty.rb]: @mdeniz done.

File: src/api/app/models/backend_package.rb (hunk 1 above)
Comment (evanrolfe) [inline src/api/app/models/backend_package.rb]: @mdeniz I've added the priority now.

File: src/api/app/jobs/project/update_packages_if_dirty_job.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class Project
+  class UpdatePackagesIfDirtyJob < ApplicationJob
```
Comment (hennevogel) [inline src/api/app/jobs/project/update_packages_if_dirty_job.rb]: Didn't we say we want all of the jobs in `app/jobs`? :-)

File: src/api/app/jobs/project/update_packages_if_dirty_job.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+    self.priority = 10
```
Comment (hennevogel) [inline src/api/app/jobs/project/update_packages_if_dirty_job.rb]: Is priority still required? I thought this is *'obsoleted'* by queues?

File: src/api/app/models/backend_package.rb [hunk 6]
```diff
@@ -20,7 +20,7 @@ class BackendPackage < ApplicationRecord
   # this is called from the UpdatePackageMetaJob and clockwork
   def self.refresh_dirty
     Package.dirty_backend_package.pluck(:project_id).uniq.each do |project_id|
-      Project.find(project_id).delay(priority: 10).update_packages_if_dirty
+      Project::UpdatePackagesIfDirtyJob.perform_later(project_id)
```
Comment (hennevogel) [inline src/api/app/models/backend_package.rb]: Why do you keep this *abstraction*? A job calls a class method of one model that calls a instance method of another class (that instantiates itself) to call an instance method on itself. What's wrong with calling `Project.find(project_id).update_packages_if_dirty.perform_later` directly?

File: src/api/config/environments/test.rb [hunk 7]
```diff
@@ -60,7 +60,9 @@
   end
 
   # TODO: This shouldn't be needed when we switch to RSpec completely
-  config.action_dispatch.rescue_responses.merge!('ActionController::InvalidAuthenticityToken' => 950 )
+  config.action_dispatch.rescue_responses['ActionController::InvalidAuthenticityToken'] = 950
+
+  config.active_job.queue_adapter = :test
```
Comment (hennevogel) [inline src/api/config/environments/test.rb]: Why these changes?

File: src/api/app/jobs/project/update_packages_if_dirty_job.rb (hunk 4 above)
Comment (evanrolfe) [inline src/api/app/jobs/project/update_packages_if_dirty_job.rb]: We did say that. What makes you think I'm not going to do this? ;)

File: src/api/app/jobs/project/update_packages_if_dirty_job.rb (hunk 5 above)
Comment (evanrolfe) [inline src/api/app/jobs/project/update_packages_if_dirty_job.rb]: It seems that priority is used to determine the priority between jobs within a queue. So the default priority is 0, so if there is a queue with priority 0 and another job with priority 10, then the 0 priority job gets completed first.

https://github.com/collectiveidea/delayed_job#parameters

File: src/api/app/models/backend_package.rb (hunk 6 above)
Comment (evanrolfe) [inline src/api/app/models/backend_package.rb]: `Project.find(project_id).update_packages_if_dirty.perform_later`

That doesn't work. I dont think active job has any equivalent to the `.delay` method that delayed jobs has. AFAIK every async job needs to be in a job class.

File: src/api/config/environments/test.rb (hunk 7 above)
Comment (evanrolfe) [inline src/api/config/environments/test.rb]: In order to use rspec to test that jobs get queued. See `spec/models/backend_package_spec.rb`

File: src/api/app/models/backend_package.rb (hunk 6 above)
Comment (hennevogel) [inline src/api/app/models/backend_package.rb]: Then at least get rid of one layer of indirection :-)

File: src/api/app/jobs/project/update_packages_if_dirty_job.rb (hunk 4 above)
Comment (hennevogel) [inline src/api/app/jobs/project/update_packages_if_dirty_job.rb]: This being a new file :-)

File: src/api/app/jobs/project/update_packages_if_dirty_job.rb (hunk 5 above)
Comment (hennevogel) [inline src/api/app/jobs/project/update_packages_if_dirty_job.rb]: k

**Comment by adrianschroeter** [comment]
@evanrolfe you would not need another table here, the only table which is checked is if a backend_package object exists for each package object. And the code is generating the missing ones in a background job.

It makes of course sense to speed up delayed_jobs, but it is IMHO better to avoid to use a second table (like delayed_jobs) at all as you proposed.

**Comment by evanrolfe** [comment]
> it would be a single select for all missing backend_package objects and an insert for each missing one without the need to handle any jobs or events at all

And how would we keep track of failures? In the way you describe that would mean that if a BackendPackage fails to be created for a Package then the job will continously retry that ad infinitum.
